src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -0,0 +1,185 @@
+  * for "read-process-write" streams, with the ability to simultaneously produce
+  * `ProducerRecord`s and commit corresponding `CommittableOffset`s using [[produce]].
+  * Records are wrapped in [[TransactionalMessage]] which allow an arbitrary value, that
+  * is a passthrough, to be included in the result.
+  */
+sealed abstract class TransactionalKafkaProducer[F[_], K, V] {
+
+  /**
+    * Produces the `ProducerRecord`s in the specified [[TransactionalMessage]]
+    * in four steps: first a transaction is initialized, then the records are
+    * placed in the buffer of the producer, then the offsets of the records are
+    * sent to the trancation, and lastly the transaction is completed. The
+    * first returned effect completes when the transaction is completed; the
+    * second effect waits for the records to have actually been sent. Note that
+    * it is very slow to wait for individual records to complete sending, but if
+    * you're sure that's what you want, then simply `flatten` the result from this
+    * function.<br>
+    * <br>
+    * If you're only interested in the passthrough value, and not the whole
+    * [[ProducerResult]], you can instead use [[producePassthrough]] which
+    * only keeps the passthrough value in the output.
+    */
+  def produce[G[+ _], P](
+    message: TransactionalMessage[F, G, K, V, P]
+  ): F[F[ProducerResult[G, K, V, P]]]
+
+  /**
+    * Like [[produce]] but only keeps the passthrough value of the
+    * [[ProducerResult]] rather than the whole [[ProducerResult]].
+    */
+  def producePassthrough[G[+ _], P](
+    message: TransactionalMessage[F, G, K, V, P]
+  ): F[F[P]]
+}
+
+private[kafka] object TransactionalKafkaProducer {
+  private[this] def createProducer[F[_], K, V](
+    settings: ProducerSettings[K, V],
+    transactionalId: String,
+    blockingEc: ExecutionContext
+  )(implicit F: Concurrent[F], C: ContextShift[F]): Resource[F, Producer[K, V]] = {
+    Resource
+      .make[F, Producer[K, V]] {
+        settings.producerFactory
+          .create(settings.withProperty(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId))
+      } { producer =>
+        F.delay {
+            producer.close(settings.closeTimeout.asJava)
+          }
+          .start
+          .flatMap(_.join)
+      }
+      .evalMap { producer =>
+        C.evalOn(blockingEc)(F.delay(producer.initTransactions())).as(producer)
+      }
+  }
+
+  def producerResource[F[_], K, V](
+    settings: ProducerSettings[K, V],
+    transactionalId: String,
+    consumerGroupId: String,
+    blockingEc: ExecutionContext
+  )(
+    implicit F: ConcurrentEffect[F],
+    C: ContextShift[F]
+  ): Resource[F, TransactionalKafkaProducer[F, K, V]] =
+    createProducer(settings, transactionalId, blockingEc).map { producer =>
+      new TransactionalKafkaProducer[F, K, V] {
+        override def produce[G[+ _], P](
+          message: TransactionalMessage[F, G, K, V, P]
+        ): F[F[ProducerResult[G, K, V, P]]] =
```

Comment:
The double-`F` wrapper isn't as useful here because (I think) committing a transaction forces a buffer flush within the producer. Is it worth keeping, or should we just flatten it out for clarity?

Reply:
Agreed, we should just return a single effect here.